### PR TITLE
Python 3.9

### PR DESCRIPTION
**Pull request by henryiii**
Rebased from #64, I don't seem to be able to push to that one for some reason.

- Add .ci_support/migrations/python39.yaml
- Remove numpy<1.19 test constraint
- MNT: Re-rendered with conda-build 3.21.4, conda-smithy 3.9.0, and conda-forge-pinning 2021.03.12.19.24.49

### CODE REVIEW

**Comment by henryiii** [comment]
> was in an excellent condition

Not really, the build number didn't get bumped. :)

**Comment by henryiii** [comment]
Oh, that's a test constraint, never mind, no build number bump needed.

**Comment by jakirkham** [comment]
That's ok. We don't need to bump the build number for this

File: recipe/meta.yaml [hunk 1]
```diff
@@ -73,9 +73,6 @@ test:
     # Need these for AOT. Do not init msvc as it may not be present
     - {{ compiler('c') }}      # [not win]
     - {{ compiler('cxx') }}    # [not win]
-    # Working around an issue with NumPy 1.19
-    # xref: https://github.com/numba/numba/issues/5929
-    - numpy >=1.15,<1.19
```
Comment (jakirkham) [inline recipe/meta.yaml:78]: Oh sorry spoke too soon. Yeah let's bump the build number if we are changing this

Comment (henryiii) [reply recipe/meta.yaml:78]: This is only a test constraint. Shouldn't affect users.

Comment (henryiii) [reply recipe/meta.yaml:78]: I though it was an install constraint, which would need a build number bump

Comment (jakirkham) [reply recipe/meta.yaml:78]: Oops 🤦‍♂️ Missed your second comment. Agree we don't need to bump the build number

**Review by jakirkham — APPROVED** [review]

File: recipe/meta.yaml (hunk 1 above)
Comment (henryiii) [reply recipe/meta.yaml:78]: We went through the exact same sequence. :)

**Comment by henryiii** [comment]
I believe there's one very minor error in test reporting in NumPy 1.19 that was discussed in https://github.com/numba/numba/issues/5929 was fixed in https://github.com/numpy/numpy/pull/17457 but never backported to the 1.19 series. Before Python 3.9, we could just pin <1.19 and avoid it. Now with Python 3.9, only 1.19 and 1.20 support it

NumPy 1.20 may have fixed this one minor test error, but it seems to have introduced 10 new errors, so that's even worse.

**Comment by henryiii** [comment]
My current thought would be to pin <1.20, and then filter the broken test. Better than filtering 10 tests.

**Comment by henryiii** [comment]
Here are the 10 failing tests for 1.20:

```
FAIL: test_nanpercentile_basic (numba.tests.test_array_reductions.TestArrayReductions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 343, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
AssertionError: when comparing [       -inf -1.70182786 -0.46359554         nan -0.53228227        -inf
 -1.41421824  0.57995536 -0.02271934         nan  0.17140244 -1.06597053
        -inf -0.77504848] and [        nan -1.70182786 -0.46359554         nan -0.53228227         nan
 -1.41421824  0.57995536 -0.02271934         nan  0.17140244 -1.06597053
         nan -0.77504848]: -inf != nan

FAIL: test_nanpercentile_basic (numba.tests.test_array_reductions.TestArrayReductions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 711, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)
AssertionError: 69 != 68

FAIL: test_nanquantile_basic (numba.tests.test_array_reductions.TestArrayReductions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 343, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
AssertionError: when comparing [       -inf -1.70182786 -0.46359554         nan -0.53228227        -inf
 -1.41421824  0.57995536 -0.02271934         nan  0.17140244 -1.06597053
        -inf -0.77504848] and [        nan -1.70182786 -0.46359554         nan -0.53228227         nan
 -1.41421824  0.57995536 -0.02271934         nan  0.17140244 -1.06597053
         nan -0.77504848]: -inf != nan


FAIL: test_nanquantile_basic (numba.tests.test_array_reductions.TestArrayReductions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 711, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)
AssertionError: 69 != 68

FAIL: test_percentile_basic (numba.tests.test_array_reductions.TestArrayReductions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 343, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
AssertionError: when comparing [-inf -inf -inf   1.] and [nan nan nan  1.]: -inf != nan

FAIL: test_percentile_basic (numba.tests.test_array_reductions.TestArrayReductions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 711, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)
AssertionError: 291 != 290

FAIL: test_quantile_basic (numba.tests.test_array_reductions.TestArrayReductions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 343, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
AssertionError: when comparing [-inf -inf -inf   1.] and [nan nan nan  1.]: -inf != nan

FAIL: test_quantile_basic (numba.tests.test_array_reductions.TestArrayReductions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 711, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)
AssertionError: 291 != 290

FAIL: test_windowing (numba.tests.test_np_functions.TestNPFunctions)
File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 343, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
AssertionError: when comparing [0.         0.18181818 0.36363636 0.54545455 0.72727273 0.90909091
 0.90909091 0.72727273 0.54545455 0.36363636 0.18181818 0.        ] and [0.         0.18181818 0.36363636 0.54545455 0.72727273 0.90909091
 0.90909091 0.72727273 0.54545455 0.36363636 0.18181818 0.        ]: 0.18181818181818177 != 0.18181818181818182 within 4.037174635000569e-17 delta (5.551115123125783e-17 difference)

FAIL: test_windowing (numba.tests.test_np_functions.TestNPFunctions)
  File "/home/conda/feedstock_root/build_artifacts/numba_1615597440943/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.9/site-packages/numba/tests/support.py", line 711, in assert_no_memory_leak
    self.assertEqual(total_alloc, total_free)
AssertionError: 12 != 11
```

**Comment by henryiii** [comment]
Or maybe 3.9 works with that limiting...

**Comment by henryiii** [comment]
Looks like it's working except for Travis, which gives "Job has been terminated due to insufficient credits balance".

**Comment by jakirkham** [comment]
Yeah that is a known issue ( https://github.com/conda-forge/status/issues/110 )

**Comment by henryiii** [comment]
I could swap out Travis for cross compilation. PyPy isn't supported yet but Numba doesn't support PyPy AFAIK. Probably will give it a day to see if the issue gets worked out?

**Comment by jakirkham** [comment]
Sure. Also ok with just merging and restarting on `master` once it clears

**Comment by henryiii** [comment]
I’d be okay with that, too

**Pull request merged**